clients/TypeScript/packages/client/src/TxSubmission/submitTx.ts

```diff
@@ -0,0 +1,127 @@
+import { nanoid } from 'nanoid'
+import { ensureSocket, InteractionContext } from '../Connection'
+import { EraMismatchError, UnknownResultError } from '../errors'
+import { errors } from './errors'
+import { baseRequest } from '../Request'
+import { EraMismatch, Ogmios, SubmitFail } from '@cardano-ogmios/schema'
+
+const isEraMismatch = (item: SubmitFail['SubmitFail']): item is EraMismatch =>
+  (item as EraMismatch).eraMismatch !== undefined
+
+export const submitTx = (bytes: string, context?: InteractionContext) => {
+  return ensureSocket<void>((socket) => {
+    return new Promise((resolve, reject) => {
+      const requestId = nanoid(5)
+      socket.once('message', (message) => {
+        const response: Ogmios['SubmitTxResponse'] = JSON.parse(message)
+        if (response.reflection.requestId !== requestId) { return }
+        if (response.methodname === 'SubmitTx') {
+          const { result } = response
+          if (result === 'SubmitSuccess') {
+            return resolve()
+          } else if ('SubmitFail' in result) {
+            const { SubmitFail } = result
+            if (isEraMismatch(SubmitFail)) {
+              const { eraMismatch } = SubmitFail
+              return reject(new EraMismatchError(eraMismatch.queryEra, eraMismatch.ledgerEra))
+            } else if (errors.byron.UtxoValidation.assert(SubmitFail)) {
+              return reject(new errors.byron.UtxoValidation.Error(SubmitFail))
+            } else if (errors.byron.TxValidation.assert(SubmitFail)) {
+              return reject(new errors.byron.TxValidation.Error(SubmitFail))
+            } else if (Array.isArray(SubmitFail)) {
+              reject(SubmitFail.map(failure => {
+                if (errors.shelley.InvalidWitnesses.assert(failure)) {
+                  return new errors.shelley.InvalidWitnesses.Error(failure)
```

Comment:
I couldn't find a more elegant way of handling these [user-defined type guards](https://levelup.gitconnected.com/user-defined-type-guards-in-typescript-fad639e4944f)